airflow/providers/sftp/hooks/sftp.py

```diff
@@ -198,7 +205,9 @@ def delete_directory(self, path: str) -> None:
         conn = self.get_conn()
         conn.rmdir(path)
 
-    def retrieve_file(self, remote_full_path: str, local_full_path: str) -> None:
+    def retrieve_file(
+        self, remote_full_path: str, local_full_path: str, regex_pattern: Optional[str] = None
+    ) -> None:
```

Comment:
One thing I really dislike about this implementation is it subtly changes semantics depending on the arguments. In this function, if `regex_pattern` is None, `remote_full_path` should point to a file, but when pattern matching is performed, it should point to a directory instead (and the file name matched with the pattern). This will be very unintuitive in practice and likely cause headaches to both users and maintainers.

Pattern matching is definitely a useful feature here, but I really don’t think this is how it should be done.